src/plugins/memdump/memdump.cpp

```diff
@@ -0,0 +1,226 @@
+    access_context_t ctx = { .translate_mechanism = VMI_TM_PROCESS_DTB, .dtb = info->regs->cr3 };
+    ctx.addr = mem_base_address_ptr;
+
+    addr_t mem_base_address;
+
+    if (VMI_SUCCESS != vmi_read_addr(vmi, &ctx, &mem_base_address))
+    {
+        PRINT_DEBUG("[MEMDUMP] Failed to read base address in NtFreeVirtualMemory\n");
+        drakvuf_release_vmi(drakvuf);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    mmvad_info_t mmvad;
+
+    if (VMI_SUCCESS != drakvuf_find_mmvad(drakvuf, info->proc_data.base_addr, mem_base_address, &mmvad))
+    {
+        PRINT_DEBUG("[MEMDUMP] Failed to find MMVAD for memory passed to NtFreeVirtualMemory\n");
+        drakvuf_release_vmi(drakvuf);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    ctx.addr = mem_base_address;
+    char magic[2];
+
+    if (VMI_SUCCESS != vmi_read_16(vmi, &ctx, (uint16_t *)&magic))
+    {
+        PRINT_DEBUG("[MEMDUMP] Failed to access memory to be used with NtFreeVirtualMemory\n");
+        drakvuf_release_vmi(drakvuf);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    if (magic[0] == 'M' && magic[1] == 'Z')
+    {
+        printf("[MEMDUMP] Interesting memory detected %d\n", info->proc_data.pid);
+        ctx.addr = mmvad.starting_vpn << 12;
+
+        size_t num_pages = mmvad.ending_vpn - mmvad.starting_vpn + 1;
+        void **access_ptrs = (void **)g_malloc(num_pages * sizeof(void *));
+
+        if (VMI_SUCCESS == vmi_mmap_guest(vmi, &ctx, num_pages, access_ptrs))
+        {
+            plugin->memdump_counter++;
+
+            char* file = nullptr;
+            if ( asprintf(&file, "%s/%d-%04d.dmp", plugin->dump_save_dir, info->proc_data.pid, plugin->memdump_counter) < 0 )
+            {
+                printf("[MEMDUMP] Failed asprintf\n");
+                drakvuf_release_vmi(drakvuf);
+                return VMI_EVENT_RESPONSE_NONE;
+            }
+
+            printf("[MEMDUMP] Writing dump with %d pages to: %s\n", (int)num_pages, file);
+
+            FILE* fp = fopen(file, "w");
+            free(file);
+            if (!fp)
+            {
+                printf("[MEMDUMP] Failed fp\n");
+                drakvuf_release_vmi(drakvuf);
+                return VMI_EVENT_RESPONSE_NONE;
+            }
+
+            for (int i = 0; i < num_pages; i++) {
+                if (access_ptrs[i]) {
+                    fwrite(access_ptrs[i], VMI_PS_4KB, 1, fp);
+                } else {
+                    uint8_t zeros[4096];
+                    fwrite(zeros, VMI_PS_4KB, 1, fp);
+                }
+            }
+
+            fclose(fp);
+            printf("[MEMDUMP] Wrote dump file\n");
+        } else {
+            printf("[MEMDUMP] Failed mmap guest\n");
+        }
+    }
+
+    drakvuf_release_vmi(drakvuf);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+memdump::memdump(drakvuf_t drakvuf, const memdump_config* c, output_format_t output)
+    : pluginex(drakvuf, output)
+{
+    this->dump_save_dir = c->dump_save_dir;
```

Comment:
Self-note: Plugin won't save any file if `--memdump-dir` would be not provided in the command line. Not sure if it should take some default or just disable plugin in such a case.

Reply:
At least in the `filedelete` plugin what we do is we provide the plugin output (that there was a file deleted) but if no folder is specified then it's not saved.

Reply:
Ok, just did similarly.